not-so-smart-contracts/substrate/weights_and_fees/README.md

```diff
@@ -0,0 +1,49 @@
+# Weights and Fees
+
+Weights and transaction fees are the two main ways to regulate the consumption of blockchain resources. The overuse of blockchain resources can allow a malicious actor to spam the network to cause a denial-of-service (DoS). Weights are used to manage the time it takes to validate the block. The larger the weight, the more "resources" / time the computation takes. Transaction fees provide an economic incentive to limit the number of resources used to perform operations; the fee for a given transaction is a function of the weight required by the transaction. 
+
+Weights can be fixed or a custom "weight annotation / function" can be implemented. A weight function can calculate the weight, for example, based on the number of database read / writes and the size of the input paramaters (e.g. a long `Vec<>`). To optimize the weight such that users do not pay too little or too much for a transaction, benchmarking can be used to empirically determine the correct weight in worst case scenarios.
+
+# Example
+In the [`pallet-bad-weights`](./pallet-bad-weights.rs) pallet, a custom weight function, `MyWeightFunction`, is used to calculate the weight for a call to `do_work`. The weight required for a call to `do_work` is `10_000_000` times the length of the `useful_amounts` vector. 
+
+```rust
+impl WeighData<(&Vec<u64>,)> for MyWeightFunction {
+    fn weigh_data(&self, (amounts,): (&Vec<u64>,)) -> Weight {
+        self.0.saturating_mul(amounts.len() as u64).into()
+    }
+}
+```
+
+However, if the length of the `useful_amounts` vector is zero, the weight to call `do_work` would be zero. A weight of zero implies that calling this function is financially cheap. This opens the opportunity for an attacker to call `do_work` a large number of times to saturate the network with malicious transactions without having to pay a large fee and could cause a DoS attack.
+
+One potential fix for this is to set a fixed weight if the length of `useful_amounts` is zero.
```

Comment:
or a static fee? Like `static fee + len * price`. This would reduce the complexity by having no branching condition

Reply:
the issue with this is that if `len == 1` then you have to pay `static_fee * 2` where i specified the weight function to be `static_fee * length`. Although the branching is not ideal, I think it represents the fix a bit more clearly. what do you think?

Reply:
I'm also leaning towards branching, even if it's not the best looking solution.

Reply:
https://github.com/crytic/building-secure-contracts/pull/124/commits/2f7851363dfd41e8e830b7d9a693c1d739261400